crates/autopilot/src/domain/settlement/mod.rs

```diff
@@ -0,0 +1,187 @@
+//! This module defines Settlement as originated from a mined transaction
+//! calldata.
+
+use {
+    crate::{
+        boundary,
+        domain::{
+            auction::{self, order},
+            eth,
+        },
+    },
+    ethcontract::{common::FunctionExt, tokens::Tokenize, U256},
+    trade::Trade,
+};
+
+mod tokenized;
+mod trade;
+
+/// Settlement originated from a calldata of a settlement transaction.
+#[derive(Debug)]
+#[allow(dead_code)]
+pub struct Settlement {
+    trades: Vec<Trade>,
+    /// Data that was appended to the regular call data of the `settle()` call
+    /// as a form of on-chain meta data. This is used to associate a
+    /// settlement with an auction.
+    auction_id: auction::Id,
+}
+
+impl Settlement {
+    /// Number of bytes that may be appended to the calldata to store an auction
+    /// id.
+    pub const META_DATA_LEN: usize = 8;
+
+    pub fn new(
+        calldata: &eth::Calldata,
+        domain_separator: &eth::DomainSeparator,
+    ) -> Result<Self, Error> {
+        let function = contracts::GPv2Settlement::raw_contract()
+            .abi
+            .function("settle")
+            .unwrap();
+        let data = calldata
+            .0
+            .strip_prefix(&function.selector())
+            .ok_or(Error::InvalidSelector)?;
+
+        let (calldata, metadata) = data.split_at(data.len() - Self::META_DATA_LEN);
+        let metadata: Option<[u8; Self::META_DATA_LEN]> = metadata.try_into().ok();
+        let auction_id = metadata
+            .map(auction::Id::from_be_bytes)
+            .ok_or(Error::MissingAuctionId)?;
+
+        let tokenized = function
+            .decode_input(calldata)
+            .map_err(|err| EncodingError::Decoding(err).with(auction_id))?;
+        let tokenized = <tokenized::Settlement>::from_token(web3::ethabi::Token::Tuple(tokenized))
+            .map_err(|err| EncodingError::Tokenizing(err).with(auction_id))?;
+
+        let (tokens, clearing_prices, decoded_trades, _interactions) = tokenized;
+
+        let mut trades = Vec::with_capacity(decoded_trades.len());
+        for trade in decoded_trades {
+            let flags = TradeFlags(trade.8);
+            let sell_token_index = trade.0.as_usize();
+            let buy_token_index = trade.1.as_usize();
+            let sell_token = tokens[sell_token_index];
+            let buy_token = tokens[buy_token_index];
+            let uniform_sell_token_index = tokens
+                .iter()
+                .position(|token| token == &sell_token)
+                .unwrap();
```

Comment:
This can cause issues in case a gas conscious solver tries to optimize the solution by getting rid of seemingly unnecessary prices.
Should we maybe have an issue tracking all the assumptions we make that need to be reevaluated once we actually move towards colocated drivers?

Reply:
This line is getting the first occurrence of the token, which is guaranteed to exist by the settlement contract right?

If you are referring to the case when the solution contains only "custom clearing prices", this code is prepared for that. Uniform and custom prices will be the same. The consequences are:
1. Gas cost will be 0 since they are a difference in between these prices (which are same in this case)
2. Protocol fee will be properly calculated as it depend solely on the custom prices. We don't need uniform prices for them.

Reply:
Yeah, was referring to the case when a solver only includes adjusted prices and no UCP. 👌